Add option to prevent users from changing custom profile fields.

tnmkr opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #22883
Completes work done in: #23852 
[CZO discussion linked in issue](https://chat.zulip.org/#narrow/stream/137-feedback/topic/First.20impression.20and.20a.20question.20.3A.29/near/1430910)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

Removed bottom margin from url inputs:
| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/133781250/d5b7f105-da4e-49ee-8f7b-3fc192132a16) | ![image](https://github.com/zulip/zulip/assets/133781250/45a0e7a0-a4ad-400f-85c9-24af7d2b496b) | 

| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/133781250/541a8aed-ca41-4b65-9063-f035a73ff02d) | ![image](https://github.com/zulip/zulip/assets/133781250/7d5bc0fb-a7f6-4cc6-a733-2da112c46d4d) |
| ![image](https://github.com/zulip/zulip/assets/133781250/1c045e76-bda3-4c6d-8f88-1cd9a3541444) | ![image](https://github.com/zulip/zulip/assets/133781250/ca549efd-f79d-425a-8117-30bb3546b160) | 

| All fields editable | All fields non editable |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/133781250/c1a73307-996b-4a0a-8027-f1b7310a9fd5) | ![image](https://github.com/zulip/zulip/assets/133781250/383af897-31a3-4fde-b35f-b67d898e8062) |
| ![image](https://github.com/zulip/zulip/assets/133781250/e34c88ba-4543-4c41-bad4-ba03a60c1041) | ![image](https://github.com/zulip/zulip/assets/133781250/b916a63a-fb57-4605-804c-187b9cd83ea2) |

Help documentation:
![image](https://github.com/zulip/zulip/assets/133781250/a0f78ad5-7921-429f-87f9-3865df8f73b0)






<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/models/custom_profile_fields.py:
```diff
@@ -79,6 +79,9 @@ class CustomProfileField(models.Model):
     display_in_profile_summary = models.BooleanField(default=False)
     required = models.BooleanField(default=False)
 
+    # Whether a non-admin user can edit this field.
```
This is imprecise -- "Whether regular users can edit this field on their own account" would be more clear. As it is, it could sound like you can edit it for any user if true.

tnmkr replied:
Modified the comment as suggested.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -10467,6 +10473,17 @@ paths:
+                    editable by users.
+
+                    This field does not affect admin users, they are always able to edit profile
+                    fields.
```
This language feels a bit off -- because the default is `true`, I feel like it's weird to talk about whether an administrator has configured it as such.

The same comment for `models.py` applies as well -- we should be clear that this applies to editing the value for their own account. And the admin sentence could be something like "Note that administrators can edit custom profile fields for any user regardless of this setting."

tnmkr replied:
This description is now:

    Whether regular users can edit this profile field on their own account.

    Note that organization administrators can edit custom profile fields for any user
    regardless of this setting.

    **Changes**: New in Zulip 9.0 (feature level 263).


Maybe this can be improved still.

timabbott commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -453,6 +453,68 @@ def test_delete_internals(self) -> None:
+        # set editable_by_user to false
+        self.login_user(iago)
+        data = {}
+        data["name"] = birthday.name
```
Is this name field required in this API query?

tnmkr replied:
No, this is no longer required. I have removed this field.

timabbott commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -888,6 +964,94 @@ def test_default_external_account_type_field(self) -> None:
+
+        # Create field with editable_by_user = false
+        data: Dict[str, Any] = {}
+        data["name"] = "Name "
```
Why the trailing whitespace here? It looks like a bug, and likely isn't what you want to test...

tnmkr replied:
Sorry, I did not think of issues due to passing unstripped whitespace to backend function.

However, as pointed above, passing `name` is no longer neccesary to modify profile fields.

timabbott commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -888,6 +964,94 @@ def test_default_external_account_type_field(self) -> None:
+        data["name"] = "Name "
+        data["editable_by_user"] = "true"
+        result = self.client_patch(f"/json/realm/profile_fields/{restricted_field_id}", info=data)
+        self.assert_json_success(result)
```
You can use `self.api_patch` to do the administrator operations without logging in as them; that often results in more readable tests.

tnmkr replied:
Modified to use `self.api_patch` for admin actions, still using `self.client_patch` for user changes though. 

Since both `/api/v1/*` and /json/*` ultimately call the same function, i assume we don't need to test both paths?

timabbott replied:
Correct!

timabbott commented on zerver/actions/custom_profile_fields.py:
```diff
@@ -187,9 +193,13 @@ def do_update_user_custom_profile_data_if_changed(
         custom_profile_field = CustomProfileField.objects.get(realm=user_profile.realm, id=field_id)
+        if not acting_user.is_realm_admin and not custom_profile_field.editable_by_user:
+            raise JsonableError(_("User does not have permission to update the field."))
         field_value = CustomProfileFieldValue.objects.get(
```
Maybe put a blank line after the check, to indicate we're moving onto a next thought.

tnmkr replied:
Added blank line as suggested.

tnmkr commented on zerver/openapi/zulip.yaml:
```diff
@@ -1855,6 +1855,7 @@ paths:
                                         "hint": "",
                                         "field_data": "",
                                         "order": 1,
+                                        "required": true,
```
I did a grep on `hint`.

tnmkr commented on web/styles/settings.css:
```diff
@@ -698,6 +698,17 @@ input[type="checkbox"] {
+    }
+
+    margin-bottom: 10px;
+}
```
Maybe it would be better to refactor all existing occurrences of `.settings_url_input` such that it no longer needs a margin?

timabbott replied:
Yeah this feels kinda hacky; @sahil839 probably has advice for what precisely to do.

sahil839 replied:
I guess we can remove the margin from `settings_url_input`, since it is anyways inconsistent with other inputs. Like `settings_text_input` also has the `margin-bottom` set as `10`, but it is overridden by other CSS which sets it to 0.
So, I guess a prep commit can be to remove this margin.

And regarding the `div`, we only really need it when we want to show a tooltip when input is disabled, and the other place where `settings_url_input`, the input element is not disabled and does not need a tooltip. And I guess removing the margin should anyways fix the tooltip without needing any extra CSS like this.

tnmkr replied:
I guess you meant `.settings_url_input` selector? It is used only in 2 places. We are removing the bottom margin for required profile fields too. Probably better to remove this as a follow-up though as it is used in org settings modal aswell.

```html
$ git grep -n ".settings_url_input"
web/styles/settings.css:741:    > .settings_url_input {
web/styles/settings.css:2114:.settings_url_input,
web/styles/settings.css:2143:    .settings_url_input,
web/templates/settings/custom_user_profile_field.hbs:29:        <input class="custom_user_field_value {{#if for_manage_user_modal}}modal_url_input{{else}}settings_url_input{{/if}}" name="{{ field.id }}" type="{{ field_type }}" value="{{ field_value.value }}" maxlength="2048" {{#unless editable_by_user}}disabled{{/unless}}/>
web/templates/settings/organization_settings_admin.hbs:146:                              name="realm_jitsi_server_url_custom_input" class="realm_jitsi_server_url_custom_input settings_url_input" maxlength="200" />
```

tnmkr replied:
Sorry, I cannot find `margin-bottom` being affected by `input[type="url"]` selector . Removing the`margin-bottom` in `.settings_url_input` completely removed this margin for me on Firefox.

sahil839 replied:
Ok. I guess I misread something. We can leave it here for now and can handle this as a follow-up.

tnmkr commented:
@sanchi-t This PR is ready for review. Please take a look when possible :)

@zulipbot add "buddy review"

tnmkr commented:
@zulipbot add "buddy review"

sanchi-t commented:
Alright, if a non-admin user has their SETTINGS/PROFILE page open and there's a custom user field marked as editable_by_user: false, and then we make that user an admin, the change doesn't immediately reflect on the page. So, the user can't edit that profile field until the page is reloaded.

I don't know if this issue is severe enough to be worked upon. You should probably wait for maintainer's review before working on a fix.

sanchi-t commented on web/styles/input_pill.css:
```diff
@@ -119,6 +119,19 @@
             perspective: 1000px;
         }
     }
+
+    &.not-editable-by-user {
+        cursor: not-allowed;
+        background-color: hsl(0deg 0% 93%);
+
+        .pill {
+            cursor: not-allowed;
+        }
+
+        .exit {
+            display: none;
+        }
+    }
```
Why are we not using `.not-editable` class? Many properties applied here are already defined there.

tnmkr replied:
This was also suggested in previous PR #23852. I agree with the reasoning provided there

> If we use not-editable class then we can avoid most of the css but would have to specify padding , border and background-color in disabled class in order to make it consistent with other types of fileds in disabled look.

sanchi-t replied:
Okay got it.

tnmkr commented on zerver/openapi/zulip.yaml:
```diff
@@ -22353,7 +22353,7 @@ components:
         data to be included in the response.
 
-        **Changes**: New in Zulip 2.1.0. Previous versions do no offer these
+        **Changes**: New in Zulip 2.1.0. Previous versions do not offer these
```
I have re-ordered this commit.

tnmkr commented on web/src/custom_profile_fields_ui.js:
```diff
@@ -32,6 +32,7 @@ export function append_custom_profile_fields(element_id, user_id) {
 
     for (const field of all_custom_fields) {
         let field_value = people.get_custom_profile_data(user_id, field.id);
+        const editable_by_user = current_user.is_admin || field.editable_by_user;
```
This name was suggested in [CZO thread](https://chat.zulip.org/#narrow/stream/137-feedback/topic/First.20impression.20and.20a.20question.20.3A.29/near/1432649)

sanchi-t replied:
That's fine then.

sanchi-t commented:
@tnmkr This PR is now ready for next round of reviews.

tnmkr commented:
@zulipbot remove "buddy review"
@zulipbot add "mentor review"

tnmkr commented:
@m-e-l-u-h-a-n Can you take a look?

timabbott commented:
I merged the 3 prep commits as the series ending with 6b255fcc147f39b4b94554e1b4e67c143fb695a9.

timabbott commented on zerver/migrations/0537_customprofilefield_editable_by_user.py:
```diff
@@ -0,0 +1,17 @@
+        migrations.AddField(
+            model_name="customprofilefield",
+            name="editable_by_user",
+            field=models.BooleanField(default=True),
```
@alexmv should we start using `db_default` for field additions like this?

alexmv replied:
I ... guess?  It makes prod inserts still work after the migration has run.

timabbott commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -942,6 +1023,84 @@ def test_default_external_account_type_field(self) -> None:
+        dummy_field_data["editable_by_user"] = "false"
+        result = self.api_post(iago, "/api/v1/realm/profile_fields", info=dummy_field_data)
+        result_json = self.assert_json_success(result)
+        restricted_field_id = result_json["id"]
```
`dummy_field_data` feels misleading as a name; maybe `realm_profile_field_data` would be appropriate?

timabbott commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -942,6 +1023,84 @@ def test_default_external_account_type_field(self) -> None:
+        # Admins can always change their own fields
+        for field_dict in iago.profile_data():
+            if field_dict["id"] == restricted_field_id:
+                self.assertEqual(field_dict["value"], None)
```
This loop[ pattern probably would benefit from having a little helper function so that these checks can look like `assert_restricted_field_value(iago, None)`

tnmkr replied:
Added this helper function. I think this what you meant?

```python
def assert_profile_field_value(self, user: UserProfile, field_id: int, field_value: str) -> None:
    for field_dict in user.profile_data():
        if field_dict["id"] == field_id:
            self.assertEqual(field_dict["value"], field_value)
```

timabbott replied:
Yep!

timabbott commented:
I posted a fairly complete set of reviews from reading the code; I've not done any manual testing, which is probably a good priority for others, especially around testing that every field type looks and renders OK if editing is disabled, which is a lot of the complexity of this PR.

tnmkr commented:
Updated to address review comments.

tnmkr commented:
Updated:
- Removed commit that moved "Profile field types" section in help docs to bottom. It is already done by ed5d291781bb52ef02e46b2f95fdf2540e3796a0.

tnmkr commented:
Marking for `maintainer review` as @m-e-l-u-h-a-n communicated is DMs that this looks good.

tnmkr commented:
@zulipbot add "maintainer review"
@zulipbot remove "mentor review"

sahil839 reviewed (commented):
Posted some comments.

I also noticed one error on client doing live-update after receivng the event a couple of times, but cannot reproduce it now. So, not sure whether it is actually a bug and if it is, whether it is in main or not.

tnmkr commented on web/styles/settings.css:
```diff
@@ -1629,6 +1629,20 @@ $option_title_width: 180px;
+
+    /* This fixes tooltip glitches. See:
+    https://github.com/zulip/zulip/pull/23852#pullrequestreview-1278592891 */
+    .not-editable-by-user-field-wrapper > .select-field-wrapper {
```
I have removed this code. Tooltip works correctly in Firefox.

tnmkr commented on web/templates/settings/custom_user_profile_field.hbs:
```diff
@@ -5,30 +5,32 @@
     <div class="alert-notification custom-field-status"></div>
     <div class="settings-profile-user-field-hint">{{ field.hint }}</div>
-    <div class="settings-profile-user-field {{#if is_empty_required_field}}empty-required-field{{/if}}">
+    <div class="settings-profile-user-field {{#if is_empty_required_field}}empty-required-field{{/if}} {{#unless editable_by_user}}not-editable-by-user-input-wrapper{{#if is_url_field}} not-editable-by-user-url-input-wrapper{{/if}}{{/unless}}">
```
Removed the `not-editable-by-user-url-input-wrapper` class.

tnmkr commented on zerver/openapi/zulip.yaml:
```diff
@@ -11064,6 +11081,16 @@ paths:
+                    Note that organization administrators can edit custom profile fields for any user
+                    regardless of this setting.
+
+                    **Changes**: New in Zulip 9.0 (feature level 280).
```
Fixed.

tnmkr commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -454,6 +454,59 @@ def test_delete_internals(self) -> None:
+        # Admins can always delete fields for themselves and other users regardless of
+        # editable_by_user. We test them deleting for their own profile here.
+        result = self.api_delete(
+            iago, "/api/v1/users/me/profile_data", {"data": orjson.dumps([birthday.id]).decode()}
```
We are testing that for `test_admin_user_can_change_profile_data` in `test_users.py`. Do you think we should have a test here aswell?

sahil839 replied:
Ok, yes we use this endpoint for changing only the custom field for the user themselves.

tnmkr commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -454,6 +454,59 @@ def test_delete_internals(self) -> None:
+        )
+        self.assert_json_success(result)
+        with self.assertRaises(CustomProfileFieldValue.DoesNotExist):
+            hamlet_biography.refresh_from_db()
```
Replaced this and other queries as suggested.

tnmkr commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -454,6 +454,59 @@ def test_delete_internals(self) -> None:
+        self.assertIsNotNone(iago_birthday.value)
+        self.assertIsNotNone(hamlet_birthday.value)
+        self.assertIsNotNone(iago_biography.value)
+        self.assertIsNotNone(hamlet_biography.value)
```
Oops! I didn't realise `values.get()` will error if no results are found. Removed these assertions.

tnmkr commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -454,6 +454,59 @@ def test_delete_internals(self) -> None:
+        iago_birthday = CustomProfileFieldValue.objects.get(user_profile=iago, field=birthday)
+        hamlet_birthday = CustomProfileFieldValue.objects.get(user_profile=hamlet, field=birthday)
+        iago_biography = CustomProfileFieldValue.objects.get(user_profile=iago, field=biography)
+        hamlet_biography = CustomProfileFieldValue.objects.get(user_profile=hamlet, field=biography)
```
Now that we are using `self.assertFalse(CustomProfileFieldValue.objects.filter(...).exists())` pattern -- these variables are not needed. Removed these.

tnmkr commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -454,6 +454,59 @@ def test_delete_internals(self) -> None:
+        realm = iago.realm
+        self.login("hamlet")
+
+        biography = CustomProfileField.objects.get(name="Biography", realm=realm)
```
Renamed.

tnmkr commented on version.py:
```diff
@@ -34,7 +34,7 @@
 # entries in the endpoint's documentation in `zulip.yaml`.
 
-API_FEATURE_LEVEL = 277  # Last bumped for Zulip 9.0
+API_FEATURE_LEVEL = 280
```
Added comment ```# Last bumped for `editable_by_user` custom profile field setting.```

sahil839 commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -454,6 +454,69 @@ def test_delete_internals(self) -> None:
+        self.assertFalse(birthday_custom_field.editable_by_user)
+        self.assertTrue(biography_custom_field.editable_by_user)
+
+        CustomProfileFieldValue.objects.get(user_profile=iago, field=birthday_custom_field)
```
The values are not stored in any variable and hence are not used. So, this can be deleted.

tnmkr replied:
Aren't these needed to assert that the value actually exists before we delete them? I removed one `get` here that is not deleted below.

tnmkr replied:
I agree. Updated to use `assertTrue` as suggested.

sahil839 commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -454,6 +454,69 @@ def test_delete_internals(self) -> None:
+        )
+
+        # Admins can always delete fields for themselves and other users regardless of
+        # editable_by_user. We test them deleting for their own profile here.
```
I see that the `DELETE` endpoint is only for user updating the field value for themselves and not others. So, might be just better to just not mention the "other users" case in the comment here as we are not testing that.

tnmkr replied:
Updated the comment here.

sahil839 commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -472,6 +535,20 @@ def test_update(self) -> None:
+        self.assertEqual(CustomProfileField.objects.count(), self.original_count)
+        result = self.client_patch(
+            f"/json/realm/profile_fields/{field.id}",
+            info={"name": "New phone number"},
```
We already test changing name below. So, can we remove this?

tnmkr replied:
Sorry, I seem to have messed up a merge conflict. Removed these lines.

sahil839 reviewed (commented):
Posted few comments. Looks good otherwise.

tnmkr commented on web/templates/settings/custom_user_profile_field.hbs:
```diff
@@ -5,30 +5,32 @@
-                <option value="{{ this.value }}" {{#if this.selected}}selected{{/if}}>{{ this.text }}</option>
-            {{/each}}
-        </select>
+        <div class="select-field-wrapper">
```
No, it isn't. I've removed this.

tnmkr commented on web/templates/settings/custom_user_profile_field.hbs:
```diff
@@ -5,30 +5,30 @@
+        <textarea maxlength="500" class="custom_user_field_value settings_textarea" name="{{ field.id }}" {{#unless editable_by_user}}disabled{{/unless}}>{{ field_value.value }}</textarea>
         {{else if is_select_field}}
-        <select class="custom_user_field_value {{#if for_manage_user_modal}}modal_select{{else}}settings_select{{/if}} bootstrap-focus-style" name="{{ field.id }}">
+        <select class="custom_user_field_value {{#if for_manage_user_modal}}modal_select{{else}}settings_select{{/if}} bootstrap-focus-style {{#unless editable_by_user}}not-editable-by-user{{/unless}}" name="{{ field.id }}" {{#unless editable_by_user}}disabled{{/unless}}>
```
No, its not. I have removed this.

sahil839 commented on web/templates/settings/custom_user_profile_field.hbs:
```diff
@@ -5,30 +5,30 @@
         {{else if is_user_field }}
-        <div class="pill-container person_picker" name="{{ field.id }}">
-            <div class="input" contenteditable="true"></div>
+        <div class="pill-container person_picker {{#unless editable_by_user}}not-editable-by-user disabled{{/unless}}" name="{{ field.id }}">
```
Is `disabled` class required here?

tnmkr replied:
Yes we need this class so we don't try setting up the typeahead for this input in `custom_profile_fields_ui.ts:146`.

sahil839 replied:
Ok.

sahil839 reviewed (commented):
Posted one comment. Looks good otherwise.

tnmkr commented:
@sahil839 I have addressed your review comment above. There are no code changes, but checks are failing on "Install dependencies" step which seems to be a [networking issue](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/main.20failing/near/1920715).

sahil839 commented on web/tests/lib/events.js:
```diff
@@ -151,6 +152,7 @@ exports.fixtures = {
                 order: 2,
                 display_in_profile_summary: false,
                 required: false,
+                editable_by_user: false,
```
These changes should be in the first commit adding the field in backend, so that `tools/check-schemas` test pass for each commit.

tnmkr replied:
Moved this change to first commit.

sahil839 reviewed (commented):
Posted one comment. You can mark it ready for integration review once that is resolved.

tnmkr commented:
@zulipbot remove "maintainer review"
@zulipbot add "integration review"

@timabbott Ready for your review

timabbott commented:
This looks good to me. I'm not entirely sure we have the perfect version of the strings; flagging the tooltip:

![image](https://github.com/user-attachments/assets/4b938cfd-6e82-4a85-954e-c19dc652cf8f)

I'm also not entirely sure about the API error message, which shouldn't be visible to users unless there's a bug or API client bug. This PR uses `"User does not have permission to update the field."`.

The prior art we have is name changes silently fail and avatar changes use this error message:

`AVATAR_CHANGES_DISABLED_ERROR = gettext_lazy("Avatar changes are disabled in this organization.")`

alya commented:
Playing with this PR, we should make this option consistent with these similar permissions:

![Screenshot 2024-08-19 at 16 09 00@2x](https://github.com/user-attachments/assets/3ff4fd49-9ad6-40f9-b940-0fced062a62b)

So we should relabel it as "Prevent users from changing this profile field", flipping the default value to unchecked.

alya commented:
We should also reorder the options, so that it's:

 Display on user card
 Required field
 Prevent users from changing this field

alya commented:
Looks good to me otherwise!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11241,6 +11258,16 @@ paths:
+                    Note that organization administrators can edit custom profile fields for any user
+                    regardless of this setting.
+
+                    **Changes**: New in Zulip 10.0 (feature level 283).
```
This number is stale.

tnmkr replied:
Fixed.

timabbott commented on web/templates/settings/edit_custom_profile_field_form.hbs:
```diff
@@ -36,6 +36,12 @@
+      setting_name="editable-by-user"
+      is_checked=editable_by_user
+      label=(t "Users can edit this field")
+      }}
```
It looks like you missed the feedback to move this setting last in the block?

tnmkr replied:
This is not ready for review yet. I only resolved the merge conflicts, I'm working on implementing the feedback now. 

Should be done in a couple of hours.

tnmkr replied:
Updated. Ready for review now.

tnmkr commented:
Adressed all feedback comments except this one:-
> So we should relabel it as "Prevent users from changing this field", flipping the default value to unchecked.

We should change the setting name internally to mirror how it's configured in the frontend. Otherwise implementing this change gets confusing as we need to maintain `true` in the backend and `false` in the frontend. 

This setting should be renamed from "editable_by_user" to something like "editable_by_admin_only"?

---
- Resolved merge conflicts, updated `zulip.yaml` and reordered migrations.
- Modified API error response text.
```diff
-            raise JsonableError(_("User does not have permission to update the field."))
+            raise JsonableError(_("You are not allowed to change this field. Contact an administrator to update it."))
```
- Reordered the checkbox.
- Updated tooltip text.
```diff
-                "Changing this field is disabled in this organization. Contact an administrator to update this field.",
+                "You are not allowed to change this field. Contact an administrator to update it.",
```

tnmkr commented:
@timabbott This is ready for review now. I also need feedback on setting name.

timabbott commented:
I think probably we should keep the API design and strings as-is; after reviewing the "Prevent ..." strings, I think a better resolution is going to be to change the existing ones eventually than do this one like them.

timabbott commented:
Merged, thanks for all the work on this @tnmkr and @Ujjawal3! @alya may want to do a quick pass on the help center and other wording before this makes it out to production.

alya commented:
Yeah, I don't think this field label is quite working for me. I started a [CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/editable.20profile.20fields/near/1947684).
